storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java

```diff
@@ -0,0 +1,417 @@
+     * @param user the user this is going to run as
+     * @param topologyId the id of the topology
+     * @throws IOException on any error
+     */
+    @SuppressWarnings("unchecked")
+    protected void writeLogMetadata(Map<String, Object> topologyConf, String user, String topologyId) throws IOException {
+        Map<String, Object> data = new HashMap<>();
+        data.put(Config.TOPOLOGY_SUBMITTER_USER, user);
+        data.put("worker-id", _workerId);
+
+        Set<String> logsGroups = new HashSet<>();
+        //for supervisor-test
+        if (topologyConf.get(Config.LOGS_GROUPS) != null) {
+            List<String> groups = (List<String>) topologyConf.get(Config.LOGS_GROUPS);
+            for (String group : groups){
+                logsGroups.add(group);
+            }
+        }
+        if (topologyConf.get(Config.TOPOLOGY_GROUPS) != null) {
+            List<String> topGroups = (List<String>) topologyConf.get(Config.TOPOLOGY_GROUPS);
+            logsGroups.addAll(topGroups);
+        }
+        data.put(Config.LOGS_GROUPS, logsGroups.toArray());
+
+        Set<String> logsUsers = new HashSet<>();
+        if (topologyConf.get(Config.LOGS_USERS) != null) {
+            List<String> logUsers = (List<String>) topologyConf.get(Config.LOGS_USERS);
+            for (String logUser : logUsers){
+                logsUsers.add(logUser);
+            }
+        }
+        if (topologyConf.get(Config.TOPOLOGY_USERS) != null) {
+            List<String> topUsers = (List<String>) topologyConf.get(Config.TOPOLOGY_USERS);
+            for (String logUser : topUsers){
+                logsUsers.add(logUser);
+            }
+        }
+        data.put(Config.LOGS_USERS, logsUsers.toArray());
+
+        File file = ConfigUtils.getLogMetaDataFile(_conf, topologyId, _port);
+        File parent = file.getParentFile();
+        if (!Utils.checkFileExists(parent)) {
+            FileUtils.forceMkdir(file.getParentFile());
+            _ops.setupStormCodeDir(ConfigUtils.readSupervisorStormConf(_conf, topologyId), file.getParentFile().getCanonicalPath());
+        }
+        Yaml yaml = new Yaml();
+        try (FileWriter writer = new FileWriter(file)) {
+            yaml.dump(data, writer);
+        }
+    }
+    
+    /**
+     * Create symlink from the containers directory/artifacts to the artifacts directory
+     * @param topologyId the id of the topology this is for
+     * @throws IOException on any error
+     */
+    protected void createArtifactsLink(String topologyId) throws IOException {
+        String workerDir = ConfigUtils.workerRoot(_conf, _workerId);
+        String topoDir = ConfigUtils.workerArtifactsRoot(_conf, topologyId);
+        if (Utils.checkFileExists(workerDir)) {
+            LOG.debug("Creating symlinks for worker-id: {} topology-id: {} to its port artifacts directory", _workerId, topologyId);
+            Utils.createSymlink(workerDir, topoDir, "artifacts", String.valueOf(_port));
+        }
```

Comment:
I agree the code way copied and pasted from before, but we should have eliminated the races so if there worker dir is not there we have a bug and should do something about it.